### ISSUE TITLE
Android TV support

### ISSUE THREAD

**Issue by MojoJojoDojo** [stenzek/duckstation#1093]
**Is your feature request related to a problem? Please describe.**
Android TV support
**Describe the solution you'd like**
Have Duckstation work over Android TV both standalone or/and libretro core.
This is also a bug report as well because the libretro core just crashes when I tried it.
**Describe alternatives you've considered**
I guess I could play it by using a PC but I rather play it over a big screen with little to no trouble.
**Additional context**
Like I said this is a bug report as well because the libretro core just crashes on my Android TV device.

**Comment by MojoJojoDojo** [comment]
Hmm the liberto core crashing is unrelated to Android TV but I guess its armv7 specific.

**Comment by stenzek** [comment]
Android armv7 libretro core seems fine for me, but slow.

**Comment by StelioKantos** [comment]
This app works great on Shield TV other than the controller. I can't find a menu to map the buttons for custom profile so, for now, when playing a game, as soon as I press any direction it just gets stuck forever in that one direction and I can't map the select button. Keep up the great work though, I'm excited for this! Thank you!

**Comment by MojoJojoDojo** [comment]
> This app works great on Shield TV other than the controller. I can't find a menu to map the buttons for custom profile so, for now, when playing a game, as soon as I press any direction it just gets stuck forever in that one direction and I can't map the select button or other buttons. The mapping isn't correct for the Shield controller. Keep up the great work though, I'm excited for this! Thank you!

How you got the games to load though?
Unless Nvidia Shield got a way to browse for files and games.

**Comment by Castrol69** [comment]
I think a part of this issue may be due to android TV not supporting SAF(storage access framework) as I'm also having issues trying to open a game on my beelink gtking which doesn't have SAF, it asks which app to use to open the file instead of going straight to SAF

**Comment by stenzek** [comment]
SAF is used for the file/directory selector. It just converts it to a native path before loading (this will change eventually).

**Comment by Castrol69** [comment]
> SAF is used for the file/directory selector. It just converts it to a native path before loading (this will change eventually).

Yep and when I try to open a file I get the message in the photo which I can't do because my box doesn't have SAF
![Uploading 20201216_235918.jpg…]()

**Comment by Castrol69** [comment]
![Uploading 20201216_235918.jpg…]()

**Comment by darkphoenixfox** [comment]
I'm glad to see that Android TV support is being considered. This is my fav PSX emulator at the moment and I would love to use it with the Shield. If I can help in any way let me know

**Comment by Castrol69** [comment]
Just thought I'd check back in and thank you for the latest update 0.1-3187.
I'm now able to manually enter the directory path ie /emulated/0/games and it works. 
I still get the "No apps can preform this action" if I try adding from the main screen but doesn't matter. 
If you have an android TV maybe try the latest version and in settings choose "edit game directories" and you may get lucky. 
Thanks again

**Comment by MojoJojoDojo** [comment]
Yeah with that directory path option you might as well call the Android version ATV compatible in a way lol.

**Comment by flatmapthatshit** [comment]
IMO this shouldn't be closed until:

1. There's a better way to add games (no "typing" of manual paths)
2. DS can be installed using the Google Play Store

Side-topic: Frontend support will follow up after an official launch for Android TB. I'm very interested in integrating DuckStation to my emulation setup (see #1847), and I even managed to hack HyperSpin's APK to launch DuckStation using it's non-standard launch intention (well, sorta, there are still crashes after exiting games and the user experience still isn't seamless).

I have to say that DuckStation is my favourite PS1 emulator. It can easily upscale to 1440p (and even 4k for some games) with PGXP enabled in a Note 10+. I would love to make it my standard option on the  NVIDIA Shield.

**Comment by stenzek** [comment]
I'm not going to be doing this any time soon. Android tv devices are way too underpowered to be worth considering.

Plus there's the whole scoped storage mess - I'm going to be moving duckstation fully over to it in the next few months as it's now a Google requirement, and I don't want to have to maintain separate builds. So if your device is missing the document selector and such, that's a problem with it, not the emu, as it's the only way to access files on Android going forward.

**Comment by MojoJojoDojo** [comment]
Don't know about underpowered I run this well on my ATV device - tried MML with 2x rendering , PGXP and Widescreen hack and it runs great.

> There's a better way to add games (no "typing" of manual paths)
I wish there was a way it remembered the manual paths I typed each time other then that I dont mind.

**Comment by aaccioly** [comment]
> Android tv devices are way too underpowered to be worth considering.

Hi @stenzek, with all due respect but I think that you are either unfamiliar or being unfair with the whole NVIDIA Tegra SoC ecosystem. It has plenty of power (Nintendo Switch runs on Tegra) and a great emulation scene. Given that DuckStation runs on my potato phone I doubt that my Shield TV would struggle with it.

> Plus there's the whole scoped storage mess - I'm going to be moving duckstation fully over to it in the next few months

Every Android software developer is currently struggling with Android 11 restrictions at the moment and I understand that we'll be talking about it until November. Nevertheless, Storage Access Framework has been present since Android 4 and API level 30 changes can be made backward compatible. DuckStation is already declared backwards compatible with API Level 23. Unless you are planning to drop support for Android Oreo soon I don't think that Android TV makes backwards compatibility significantly arder. 

> I don't want to have to maintain separate builds

I'm not sure what you mean here, but you don't need to maintain separate builds. There are plenty of emulators out there using Android App Bundle and delivering the same arm based builds for Mobile and Android TV. 
As long as both leanback and touchscreen support are declared optional in the manifest file you can [opt-in](https://developer.android.com/distribute/best-practices/launch/distribute-tv#opt_in) to Android TV. 
I haven't tested DuckStation on Android TV myself, but by what @flatmapthatshit and @MojoJojoDojo are saying it looks like DuckStation APK is almost there. Other than the file picker stuff (see bellow) it sounds like DuckStation can get away with just a few changes to the manifest file and a couple of TV Banners - well, maybe some cosmetic changed depending on how picky the reviewer assigned to your app is (see [TV apps Guidelines](https://developer.android.com/docs/quality-guidelines/tv-app-quality) ).

> So if your device is missing the document selector and such, that's a problem with it, not the emu

 [`ACTION_OPEN_DOCUMENT_TREE`](https://developer.android.com/reference/android/content)/Intent#ACTION_OPEN_DOCUMENT_TREE) works fine in my NVIDIA Shield. I totally agree that you shouldn't go out of your way to implement support for weak devices that don't implement standard API features (I'm all too familiar with cheap Android TV boxes that can't display a file picker); but I can assure you that, whatever is happening, the problem here isn't a lack of support for SAF.

Just to make my intentions completely clear. I'm not by any way demanding support for Android TV or saying that such an undertaking is easy. It's your project, you know it's priorities better than anyone. I'm just trying to say that NVIDIA Shield is a completely different category of device than the typical cheap Android TV box. It's a great device for emulation and it would be awesome if DuckStation could officially support it one day. I'm sure that there's a very enthusiastic community of NVIDIA Shield users that would agree with me.

**Comment by MojoJojoDojo** [comment]
You know considering I got my device from my TV provider company (its not a shield tv) it does the job well..
I think that devices released from the last years (around 2019-2021) like mine could perform well enough for the most time.
Just noting my personal experience.

**Comment by flatmapthatshit** [comment]
Android TV can be anything really. From underpowered x86_64 boxes all the way to high end arm devices with custom GPUs. Support for ARMv7 and AArch64 would be enough for 99% of Android TV users. And yes, there are many devices with enough horsepower to emualte 5th Gen Consoles.

**Comment by stenzek** [comment]
Most of the devices I've seen are very weak (Cortex-A53, clocked below 2ghz, 1GB RAM, not necessarily even a GLES3 GPU).

> Hi @stenzek, with all due respect but I think that you are either unfamiliar with or being unfair to the whole NVIDIA Tegra SoC ecosystem.

The Tegra is an 8+ year old CPU design on a nearly just as old GPU. Your phone likely has more compute power, unless it's very low end.

> Every Android software developer is currently struggling with Android 11 restrictions at the moment..

DuckStation is not "struggling" here - I have a SAF based scanner on a branch ready for merging, just need to add options to let the user specify their save directory.

> I'm not sure what you mean here, but you don't need to maintain separate builds for Android TV.

The issue is most Android TV devices do not provide a document selector. I'm not going to go to the effort of creating a custom picker just for a handful of devices which aren't powerful enough anyway.

Therefore, it's actually a really shitty situation - you can't request legacy storage anymore after November, but the ATV devices don't provide any other way to access files? So basically you can't publish ATV apps which require storage anymore...

I also don't own any Android TV devices to develop/test on - do you seriously expect me to run out and buy an overpriced STV when the project brings in zero income and actually costs me money out of pocket to run...

tldr; it's not something which interests me at all, hardware is underpowered. If someone else wants to implement it, I'm happy to review/merge, but it's not something I'm going to be doing.

**Comment by flatmapthatshit** [comment]
@stenzek, everyone here is telling you that their Android TV is capable of running your emulator. 
Tegra X1+ was released in 2019, while it's indeed an old architecture, the new devices  support GLS 3.2 just fine. Sure, ARM Cortex-A57 isn't exactly a top CPU for today's standards, but it doesn't need to be. Here's an OpenGL benchmark comparing an Nvidia shield with Samsung's current flagship: https://gfxbench.com/compare.jsp?benchmark=gfx50&did1=26084812&os1=Android&api1=gl&hwtype1=GPU&hwname1=NVIDIA%28R%29+Tegra%28R%29+X1&D2=Samsung+Galaxy+S20+Ultra+5G+%28SM-G988U1%29

You have already implemented an alternative way to add ROMs for devices without a document selector (although I agree with @MojoJojoDojo that persisting the last entered path would certainly improve usability).

I would be very happy to donate a Shield or bounty $200 towards implementing Android TV support.

@aaccioly, you own a shield right? Would you like to get involved?

**Comment by stenzek** [comment]
I never said it couldn't run it. I said I don't want to develop a frontend just for one device which I don't own, nor have any interest in. You're asking me to spend hours/days on work which I don't want to do, on a project which I'm already giving away for free. If DuckStation was bringing in money and you were a customer, sure, but this is a hobby for me and I'll only continue to work on it so long as I find it interesting.

> You have already implemented an alternative way to add ROMs for devices without a document selector (although I agree with @MojoJojoDojo that persisting the last entered path would certainly improve usability).

This is only temporary and will be removed once I merge the SAF scanner. Only `content://` URIs will be scannable, as per Google requirements. And I'm not keeping a second build around with a legacy scanner which I can't publish on the store.

**Comment by aaccioly** [comment]
First of all, Happy Easter.

> @aaccioly, you own a shield right? Would you like to get involved?

NO :D. 
Ok. Let me try to help by breaking down what you are requesting and what stenzek is saying.

>  DS can be installed using the Google Play Store.

The main developer made it abundantly clear that he isn't interested in supporting Android TV any time soon. Even if someone takes the time to make the necessary code changes, there's much more to it than a one-off contribution:

1. Bureocracy to get it published will inevitably require stenzek's attention.
2. Android API is a moving target. Google keeps restricting   what we can do and randomly taking down emulators from the store. Supporting Android (not only Android TV but mobile as well) is a burden and a long term commitment that I'm not willing to do.

As a developer I understand the frustration involved in dealing with Google and Apple. I'm ok with paying money for Android emulators and reward the good work of developers willing to put up with it. But I'm not ok with maintaining a fork by myself or pushing someone's else hobby project in a direction that the author is not willing to go.

Honestly, if you are serious about putting a bounty on something, libretro guys would probably be more responsive to it. If I'm not mistaken there's a libretro core available for Duckstation right? RetroArch has pretty good support for Android TV. Otherwise I'm sorry to say but for now we are stuck with sideloading the APK.

> There's a better way to add games (no "typing" of manual paths)

Let's break it down this further:

*NVIDIA Shield and Android TV devices that support SAF:* I honestly have no idea why the functionality to add a directory isn't working as is. It should. Maybe when stenzek finishes merging Android 11 support it will work. If it doesn't I'll have a look.

*All other Android TV platforms:* as mentioned by stenzek, it would require a separate build. Again, assuming that Duckstation's libretro core works well, just go with RetroArch.
If RetroArch does not work with your platform, you will unfortunately have to fork the project (again, I'm not willing to do it myself for the reasons stated above).

Nevertheless, it is a simple enough change to hack.

First a integrate a custom file picker such as this one to the project: https://github.com/spacecowboy/NoNonsense-FilePicker (instructions are pretty easy to follow).

Then proceed  to hack DuckStation's [`MainActivity`](https://github.com/stenzek/duckstation/blob/c5e14c2f9f96130fb1e24b4a550248c8115627f0/android/app/src/main/java/com/github/stenzek/duckstation/MainActivity.java#L272-L281).

In `startAddGameDirectory`:

1. Add an if statement to check if the app is running on an Android TV device (see https://stackoverflow.com/a/30059511/664577)
2.  If it is a TV use your choice of file picker (see: https://github.com/spacecowboy/NoNonsense-FilePicker#starting-the-picker-in-your-app). Make sure to set it to directory mode.
3.  If it isn't a TV just use the [current code](https://github.com/stenzek/duckstation/blob/c5e14c2f9f96130fb1e24b4a550248c8115627f0/android/app/src/main/java/com/github/stenzek/duckstation/MainActivity.java#L195-L199)

You will also need to change  `onActivityResult`:

Extract a string from the path and forward it to `GameDirectoriesActivity`. [Current code](https://github.com/stenzek/duckstation/blob/c5e14c2f9f96130fb1e24b4a550248c8115627f0/android/app/src/main/java/com/github/stenzek/duckstation/MainActivity.java#L272-L281), [Handling the result with FilePicker](https://github.com/spacecowboy/NoNonsense-FilePicker#handling-the-result).

For the sake of consistentcy, make the same changes in [`GameDirectoriesActivity`](https://github.com/stenzek/duckstation/blob/c5e14c2f9f96130fb1e24b4a550248c8115627f0/android/app/src/main/java/com/github/stenzek/duckstation/GameDirectoriesActivity.java#L288-L334).

That's it. It will work fine for API Level <= 29 (Android 11 boxes will have to provide support for SAF anyway, otherwise they are going to be useless).

**Comment by stenzek** [comment]
I'd rather not maintain multiple paths depending on API version - if Google wants to force SAF to be a requirement, then it's not our responsibility to provide fallbacks. It's just dead code which is useless on probably 95+% of devices.

As far as I'm aware, the current solution works fine on Shield TVs. So I really don't see what the big deal is - sideload it, and it'll work fine. Nobody's ever actually described what the issue is.

**Comment by MojoJojoDojo** [comment]
Hrydgard from PPSSPP contacted Google about scoped storage on Android TV : 
https://issuetracker.google.com/issues/169471812
So maybe it will be fixed or updated... who knows.

**Comment by aaccioly** [comment]
> Hrydgard from PPSSPP contacted Google about scoped storage on Android TV :
> https://issuetracker.google.com/issues/169471812
> [hrydgard/ppsspp#11997 (comment)](https://github.com/hrydgard/ppsspp/issues/11997#issuecomment-815806301)
> So maybe it will be fixed or updated... who knows.

I wouldn't get my hopes up (from past experience with Google). They will blame device manufacturers for not supporting the API. I would prepare myself to side-load some apps and completely ditch others.

**Comment by stenzek** [comment]
Indeed. I'll probably close this request once I migrate user data over to scoped storage, I moved scanning over today. After all the changes and to meet Google requirements, duckstation won't request external storage permissions at all, so a fallback path won't be usable either. Any ATV devices with storage access framework support will continue to work fine.

**Comment by stenzek** [comment]
Closing since I have no plans to write a separate interface for Android TV, since I'm not interested in it, and have no ATV devices. Sideloading the current app should work on powerful-enough devices.
